oidc.go

```diff
@@ -238,6 +242,73 @@ func (p *Provider) UserInfo(ctx context.Context, tokenSource oauth2.TokenSource)
 	return &userInfo, nil
 }
 
+// EndSessionOptions is optional parameters for the Initiated Logout by https://openid.net/specs/openid-connect-session-1_0.html#RPLogout
+type EndSessionOptions struct {
+	IDTokenHint string
+	RedirectURL string
+	State       string
+}
+
+// EndSession uses the token source to query the provider's end session endpoint.
+func (p *Provider) EndSession(ctx context.Context, tokenSource oauth2.TokenSource, options EndSessionOptions) error {
+	endSessionURL, err := makeEndSessionEndpoint(p.endSessionURL, options)
+	if err != nil {
+		return fmt.Errorf("oidc: error of make an end session url: %v", err)
+	}
+
+	req, err := http.NewRequest(http.MethodGet, endSessionURL, nil)
+	if err != nil {
+		return fmt.Errorf("oidc: create GET request: %v", err)
+	}
+
+	token, err := tokenSource.Token()
+	if err != nil {
+		return fmt.Errorf("oidc: get access token: %v", err)
+	}
+
+	token.SetAuthHeader(req)
+
+	resp, err := doRequest(ctx, req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	return nil
+}
+
+func makeEndSessionEndpoint(endSessionURL string, options EndSessionOptions) (endURL string, err error) {
```

Comment:
nit: rename "options" to "o"
nit: rename "makeEndSessionEndpoint" to "endSessionEndpoint"

Also please use shorter variable names

```go
	if endSessionURL == "" {
		return "", errors.New("oidc: end session endpoint is not supported by this provider")
	}

	u, err := url.Parse(endSessionURL)
	if err != nil {
		return "", fmt.Errorf("oidc: wrong format of end session endpoint %s: %v", p.endSessionURL, err)
	}
	q := u.Query()
	if o.IDTokenHint != "" {
		q.Set("id_token_hint", o.IDTokenHint)
	}
	if o.RedirectURL != "" {
		q.Set("post_logout_redirect_uri", o.RedirectURL)
	}
	if o.State != "" {
		q.Set("state", o.State)
	}
	u.RawQuery = q.Encode()
	return u.String(), nil
```

Reply:
Also, please add a test for this method.